forgerock-openbanking-uk-aspsp-as/forgerock-openbanking-uk-aspsp-as-gateway/forgerock-openbanking-uk-aspsp-as-gateway-server/src/main/java/com/forgerock/openbanking/aspsp/as/api/accesstoken/AccessTokenApiController.java

```diff
@@ -23,82 +23,83 @@
         this.amGateway = amGateway;
         this.headLessAccessTokenService = headLessAccessTokenService;
-        this.tppStoreService = tppStoreService;
         this.amGatewayService = amGatewayService;
         this.jwtOverridingService = jwtOverridingService;
-        this.tokenUsageService = tokenUsageService;
+        this.matlsRequestVerificationService = matlsRequestVerificationService;
     }
 
     @Override
     public ResponseEntity getAccessToken(MultiValueMap paramMap, String authorization, Principal principal,
                                          HttpServletRequest request) throws OBErrorResponseException, OBErrorException {
-
-        PairClientIDAuthMethod clientIDAuthMethod = verifyMATLSMatchesRequest(paramMap, authorization, principal);
+        log.debug("getAccessToken(), paramMap {}", paramMap);
+        PairClientIDAuthMethod clientIDAuthMethod =
+                matlsRequestVerificationService.verifyMATLSMatchesRequest(paramMap, authorization, principal);
 
         AMGateway amGateway = this.amGateway;
         //The token endpoint can also be used as audience, as per OIDC spec
-        if (clientIDAuthMethod.authMethod == OIDCConstants.TokenEndpointAuthMethods.PRIVATE_KEY_JWT) {
-            amGateway = amGatewayService.getAmGateway((String) paramMap.getFirst(CLIENT_ASSERTION));
+        if (clientIDAuthMethod.getAuthMethod() == OIDCConstants.TokenEndpointAuthMethods.PRIVATE_KEY_JWT) {
+            String  clientAssertion = (String) paramMap.getFirst(OIDCConstants.OIDCClaim.CLIENT_ASSERTION);
+            if(clientAssertion == null || clientAssertion.isBlank() || clientAssertion.isEmpty()){
+                log.debug("getAccessToken() clientAssertion was null or blank");
+                throw new OBErrorResponseException(
+                        OBRIErrorType.ACCESS_TOKEN_INVALID.getHttpStatus(),
+                        OBRIErrorResponseCategory.ACCESS_TOKEN,
+                        OBRIErrorType.ACCESS_TOKEN_INVALID.toOBError1("No client_assertion in body"
+                ));
+            }
+            amGateway = amGatewayService.getAmGateway(clientAssertion);
         }
+
         // can throw a UnsupportedOIDCGrantTypeException
-        OIDCConstants.GrantType grantType = OIDCConstants.GrantType.fromType((String) paramMap.getFirst(GRANT_TYPE));
-        ResponseEntity responseEntity;
+        OIDCConstants.GrantType grantType =
+                OIDCConstants.GrantType.fromType((String) paramMap.getFirst(OIDCConstants.OIDCClaim.GRANT_TYPE));
+        ResponseEntity<AccessTokenResponse> responseEntity;
+
         switch (grantType) {
             case HEADLESS_AUTH:
-                responseEntity = headLessAccessTokenService.getAccessToken(amGateway, clientIDAuthMethod, paramMap, request);
+                responseEntity = headLessAccessTokenService.getAccessToken(amGateway, clientIDAuthMethod, paramMap,
```

Comment:
No tests for this path. Could do with a tests for the headless auth path, and tests for the headlessAccessTokenService

Reply:
Have added a test for this path and ensured that when the responseEntity status code is not success or redirection then we return an OBErrorResponseException.